### PR TITLE
[core] Implement 'in' expression.

### LINKED ISSUE #15893: Implement 'in' expression
https://github.com/mapbox/mapbox-gl-js/pull/8876

### PR DESCRIPTION
**Pull request by Chaoba**
## Launch Checklist
Resolves #15893 
This PR implements `in` expression and turns on the corresponding tests in `expression-test`.

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] apply `needs changelog` label if a changelog is needed (remove label when added)

### CODE REVIEW

File: include/mbgl/style/expression/in.hpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include <memory>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+class In : public Expression {
+public:
+    In(std::unique_ptr<Expression> needle_, std::unique_ptr<Expression> haystack_)
+        : Expression(Kind::In, type::Boolean), needle(std::move(needle_)), haystack(std::move(haystack_)) {}
```
Comment (alexshalamov) [inline include/mbgl/style/expression/in.hpp]: could you add asserts for needle and haystack

File: include/mbgl/style/expression/in.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <mbgl/style/conversion.hpp>
+#include <mbgl/style/expression/expression.hpp>
+#include <memory>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+class In : public Expression {
```
Comment (alexshalamov) [inline include/mbgl/style/expression/in.hpp]: `class In final : ...`

File: include/mbgl/style/expression/in.hpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    bool operator==(const Expression& e) const override {
+        if (e.getKind() == Kind::In) {
+            auto rhs = static_cast<const In*>(&e);
+            return *needle == *(rhs->needle) && *haystack == *(rhs->haystack);
+        }
+        return false;
+    }
+
+    std::vector<optional<Value>> possibleOutputs() const override { return {nullopt}; }
```
Comment (alexshalamov) [inline include/mbgl/style/expression/in.hpp]: In expression can return either true or false, regardless of a provided context / input.
`return {{ true }, { false }};`

File: src/mbgl/style/expression/in.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+#include <mbgl/style/expression/type.hpp>
+#include <mbgl/util/string.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+EvaluationResult In::evaluate(const EvaluationContext& params) const {
+    const EvaluationResult evaluatedNeedle = needle->evaluate(params);
+    const EvaluationResult evaluatedHeystack = haystack->evaluate(params);
```
Comment (alexshalamov) [inline src/mbgl/style/expression/in.cpp]: evaluation of haystack can be moved after `if (!evaluatedNeedle) {`

File: src/mbgl/style/expression/in.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        const auto heystackString = evaluatedHeystack->get<std::string>();
+        std::string needleValue = "";
+        if (evaluatedNeedleType == type::Boolean) {
+            needleValue = evaluatedNeedle->get<bool>() ? "true" : " false";
+        } else if (evaluatedNeedleType == type::String) {
+            needleValue = evaluatedNeedle->get<std::string>();
+        } else if (evaluatedNeedleType == type::Number) {
+            needleValue = std::to_string(evaluatedNeedle->get<double>());
+            needleValue.erase(needleValue.find_last_not_of('0') + 1, std::string::npos);
+            needleValue.erase(needleValue.find_last_not_of('.') + 1, std::string::npos);
```
Comment (alexshalamov) [inline src/mbgl/style/expression/in.cpp]: Could you explain why this is required.

File: src/mbgl/style/expression/in.cpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        }
+        return EvaluationResult(heystackString.find(needleValue) != std::string::npos);
+    } else {
+        const auto heystackArray = evaluatedHeystack->get<std::vector<Value>>();
```
Comment (alexshalamov) [inline src/mbgl/style/expression/in.cpp]: I think this will crash if `evaluatedHeystack` is `Null`. Could you add unit tests?

File: src/mbgl/style/expression/in.cpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        if (evaluatedNeedleType == type::Boolean) {
+            auto needleValue = evaluatedNeedle->get<bool>();
+            result = find(heystackArray.begin(), heystackArray.end(), needleValue) != heystackArray.end();
+        } else if (evaluatedNeedleType == type::String) {
+            auto needleValue = evaluatedNeedle->get<std::string>();
+            result = find(heystackArray.begin(), heystackArray.end(), needleValue) != heystackArray.end();
+        } else if (evaluatedNeedleType == type::Number) {
+            auto needleValue = evaluatedNeedle->get<double>();
+            result = find(heystackArray.begin(), heystackArray.end(), needleValue) != heystackArray.end();
+        }
```
Comment (alexshalamov) [inline src/mbgl/style/expression/in.cpp]: Value has comparison operator. Needle type comparability is checked on[ line 23 `isComparableType`](https://github.com/mapbox/mapbox-gl-native/pull/16162/files#diff-5d51b639578982530d698f5b5fd0536eR23).

Whole block can be rewritten as:

```c++
return EvaluationResult(find(heystackArray.begin(), heystackArray.end(), evaluatedNeedle) != heystackArray.end());
```

File: include/mbgl/style/expression/in.hpp [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+    std::string getOperator() const override { return "in"; }
+
+private:
+    std::unique_ptr<Expression> needle;
+    std::unique_ptr<Expression> haystack;
+    static bool isComparableType(type::Type type);
+    static bool isComparableRuntimeValue(type::Type type);
+    static bool isSearchableRuntimeValue(type::Type type);
```
Comment (alexshalamov) [inline include/mbgl/style/expression/in.hpp]: these can be moved to anonymous namespace within cpp file.

File: src/mbgl/style/expression/in.cpp (hunk 5 above)
Comment (Chaoba) [reply src/mbgl/style/expression/in.cpp]: Since we use `evaluatedNeedle->get<double>()` to get a double value, it will has 6 tail 0. For example, if needle is 6, `needleValue` will be "6.000000", which will cause find failed.

File: src/mbgl/style/expression/in.cpp (hunk 6 above)
Comment (Chaoba) [reply src/mbgl/style/expression/in.cpp]: We have checked Null in `isSearchableRuntimeValue` in line 29 to insure it will not happens here.

Comment (Chaoba) [reply src/mbgl/style/expression/in.cpp]: What's more, we already have test case that Haystack is null like [assert-string](https://github.com/mapbox/mapbox-gl-js/blob/c82a0b8c89bb964824996e25f54c31cfaf419e61/test/integration/expression-tests/in/assert-string/test.json)

File: src/mbgl/style/expression/in.cpp (hunk 7 above)
Comment (Chaoba) [reply src/mbgl/style/expression/in.cpp]: The type of `evaluatedNeedle` is `EvaluationResult ` , it can't be compared with type `Value`.

File: include/mbgl/style/expression/in.hpp (hunk 1 above)
Comment (Chaoba) [reply include/mbgl/style/expression/in.hpp]: We have checks in `evaluate` and `parse` method. If add asserts here it will be duplicate.

File: src/mbgl/style/expression/in.cpp (hunk 7 above)
Comment (alexshalamov) [reply src/mbgl/style/expression/in.cpp]:
``` c++
    } else {
        const auto haystackArray = evaluatedHaystack->get<std::vector<Value>>();
        return EvaluationResult(std::find(haystackArray.begin(), haystackArray.end(), *evaluatedNeedle) != haystackArray.end());
    }
```

File: include/mbgl/style/expression/in.hpp (hunk 1 above)
Comment (alexshalamov) [reply include/mbgl/style/expression/in.hpp]: When someone would add `in` expression to dsl, constructor might be called directly, without parse. So, would be good to have guards in constructor

File: src/mbgl/style/expression/in.cpp (hunk 6 above)
Comment (alexshalamov) [reply src/mbgl/style/expression/in.cpp]: > We have checked Null in isSearchableRuntimeValue in line 29 to insure it will not happens here.

Why isSearchableRuntimeValue returns true for Null type? And then, evaluation returns if it is Null type in https://github.com/mapbox/mapbox-gl-native/pull/16162/files#diff-5d51b639578982530d698f5b5fd0536eR49

File: src/mbgl/style/expression/in.cpp [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+}
+}
+
+EvaluationResult In::evaluate(const EvaluationContext& params) const {
+    const EvaluationResult evaluatedNeedle = needle->evaluate(params);
+    if (!evaluatedNeedle) {
+        return evaluatedNeedle.error();
+    }
+
+    const EvaluationResult evaluatedHaystack = haystack->evaluate(params);
```
Comment (alexshalamov) [inline src/mbgl/style/expression/in.cpp]: If haystack may be a nullset, it is better to evaluate it first and then `return EvaluationResult(false)` right away.

**Comment by alexshalamov** [comment]
@Chaoba could you also add equality tests for "in" expression to `test/fixtures/expression_equality/`?

File: include/mbgl/style/expression/in.hpp (hunk 1 above)
Comment (Chaoba) [reply include/mbgl/style/expression/in.hpp]: ok

File: src/mbgl/style/expression/in.cpp [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        if (evaluatedNeedleType == type::Boolean) {
+            needleValue = evaluatedNeedle->get<bool>() ? "true" : " false";
+        } else if (evaluatedNeedleType == type::String) {
+            needleValue = evaluatedNeedle->get<std::string>();
+        } else if (evaluatedNeedleType == type::Number) {
+            needleValue = std::to_string(evaluatedNeedle->get<double>());
+            needleValue.erase(needleValue.find_last_not_of('0') + 1, std::string::npos);
+            needleValue.erase(needleValue.find_last_not_of('.') + 1, std::string::npos);
+        }
+        return EvaluationResult(haystackString.find(needleValue) != std::string::npos);
```
Comment (alexshalamov) [inline src/mbgl/style/expression/in.cpp]: I think this whole block can be rewritten as:

``` c++
const auto haystackString = evaluatedHaystack->get<std::string>();
const auto needleString = toString(evaluatedNeedle);
return EvaluationResult(haystackString.find(needleString) != std::string::npos);
```

File: include/mbgl/style/expression/in.hpp [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    EvaluationResult evaluate(const EvaluationContext& params) const override;
+    void eachChild(const std::function<void(const Expression&)>&) const override;
+
+    bool operator==(const Expression& e) const override {
+        if (e.getKind() == Kind::In) {
+            auto rhs = static_cast<const In*>(&e);
+            return *needle == *(rhs->needle) && *haystack == *(rhs->haystack);
+        }
+        return false;
+    }
```
Comment (alexshalamov) [inline include/mbgl/style/expression/in.hpp]: move to cpp

File: include/mbgl/style/expression/in.hpp [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+    std::vector<optional<Value>> possibleOutputs() const override { return {{true}, {false}}; }
```
Comment (alexshalamov) [inline include/mbgl/style/expression/in.hpp]: move to cpp

File: src/mbgl/style/expression/in.cpp [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        return ParseResult();
+    }
+
+    ParseResult needle = ctx.parse(arrayMember(value, 1), 1, {type::Value});
+    ParseResult haystack = ctx.parse(arrayMember(value, 2), 2, {type::Value});
+
+    if (!needle || !haystack) return ParseResult();
+
+    type::Type needleType = (*needle)->getType();
+    type::Type haystackType = (*haystack)->getType();
```
Comment (alexshalamov) [inline src/mbgl/style/expression/in.cpp]: unused?

File: src/mbgl/style/expression/in.cpp [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+ParseResult In::parse(const Convertible& value, ParsingContext& ctx) {
+    assert(isArray(value));
+
+    std::size_t length = arrayLength(value);
+    if (length != 3) {
+        ctx.error("Expected 2 arguments, but found " + util::toString(length - 1) + " instead.");
+        return ParseResult();
+    }
+
+    ParseResult needle = ctx.parse(arrayMember(value, 1), 1, {type::Value});
```
Comment (alexshalamov) [inline src/mbgl/style/expression/in.cpp]: Maybe return early? Same for haystack.
```c++
ParseResult needle = ctx.parse(arrayMember(value, 1), 1, {type::Value});
if (!needle) return ParseResult();
```

File: src/mbgl/style/expression/in.cpp (hunk 10 above)
Comment (Chaoba) [reply src/mbgl/style/expression/in.cpp]: Awesome!

File: CMakeLists.txt [hunk 15]
```diff
@@ -566,6 +566,7 @@ add_library(
     ${PROJECT_SOURCE_DIR}/src/mbgl/style/expression/get_covering_stops.cpp
     ${PROJECT_SOURCE_DIR}/src/mbgl/style/expression/image.cpp
     ${PROJECT_SOURCE_DIR}/src/mbgl/style/expression/image_expression.cpp
+    ${PROJECT_SOURCE_DIR}/src/mbgl/style/expression/in.cpp
     ${PROJECT_SOURCE_DIR}/src/mbgl/style/expression/interpolate.cpp
```
Comment (zmiao) [inline CMakeLists.txt]: Add in.hpp header in cmake file also.

Comment (Chaoba) [reply CMakeLists.txt]: Good catch 👍

**Comment by Chaoba** [comment]
Waiting for [https://github.com/mapbox/mapbox-gl-js/pull/9295](https://github.com/mapbox/mapbox-gl-js/pull/9295) , as there is a failed test case blocks now.

File: CHANGELOG.md [hunk 16]
```diff
@@ -22,10 +22,14 @@
    
   The `within expression` enables checking whether a feature is inside a pre-defined geometry set/boundary or not. This `within expression` returns a boolean value, `true` indicates that the feature being evaluated is inside the geometry set. The returned value can be then consumed as input by another expression or used directly by a paint/layer property.
  
-  Support for using `within expression` with layout propery will be implemented separately.
+  Support for using `within expression` with layout property will be implemented separately.
 
-- [core] Add support for using `within expression` with layout propery. ([#16194](https://github.com/mapbox/mapbox-gl-native/pull/16194))
+- [core] Add support for using `within expression` with layout property. ([#16194](https://github.com/mapbox/mapbox-gl-native/pull/16194))
```
Comment (zmiao) [inline CHANGELOG.md]: Thanks 🙇‍♀

File: CHANGELOG.md [hunk 17]
```diff
@@ -22,10 +22,14 @@
...
 
+- [core] Add support for `in expression`. ([#16162](https://github.com/mapbox/mapbox-gl-native/pull/16162))
+   
+  The `in expression` enables checking whether a Number/String/Boolean type needle is in a String/Array and returns a boolean value.
```
Comment (ryanhamley) [inline CHANGELOG.md]: It might be confusing to use the term "needle" here, at least without also using the corresponding "haystack" to make the analogy clear. I'd also probably put them in quotation marks to make it clear that these are terms of art and not specific technical terms.

**Comment by alexshalamov** [comment]
@Chaoba FYI https://github.com/mapbox/mapbox-gl-js/pull/9295

**Comment by tmpsantos** [comment]
@Chaoba thanks a lot for your contribution.
